internal/devpkg/narinfo_cache.go

```diff
@@ -125,17 +102,21 @@ func FillNarInfoCache(ctx context.Context, packages ...*Package) error {
 	return group.Wait()
 }
 
-// fillNarInfoCache fills the cache value for the narinfo of this package,
+// fillNarInfoCacheIfNeeded fills the cache value for the narinfo of this package,
 // assuming it is eligible for the binary cache. Callers are responsible
 // for checking isEligibleForBinaryCache before calling this function.
 //
 // NOTE: this must be concurrency safe.
-func (p *Package) fillNarInfoCache() error {
+func (p *Package) fillNarInfoCacheIfNeeded() (bool, error) {
+	// This check if fine to do without a lock, because we never remove/replace values
+	if status, alreadySet := isNarInfoInCache.status[p.Raw]; alreadySet {
+		return status, nil
+	}
```

Comment:
I don't think this is correct:

https://go.dev/doc/faq#atomic_maps
> Map access is unsafe only when updates are occurring. As long as all goroutines are only reading—looking up elements in the map, including iterating through it using a for range loop—and not changing the map by _assigning to elements_  or doing deletions, it is safe for them to access the map concurrently without synchronization.

https://go.dev/blog/maps
> [Maps are not safe for concurrent use](https://go.dev/doc/faq#atomic_maps): it’s not defined what happens when you read and write to them simultaneously. If you need to read from and write to a map from concurrently executing goroutines, the accesses must be mediated by some kind of synchronization mechanism. One common way to protect maps is with [sync.RWMutex](https://go.dev/pkg/sync/#RWMutex).

Reply:
It also makes sense to me that its not safe, because writing to a map can involve needing to allocate more memory and potentially move elements around...